src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+            CopyToBuffer(result, evt);
+
+            input.Advance(length);
+
+            ValueTask<FlushResult> flushTask = input.FlushAsync();
+
+            if (!flushTask.IsCompletedSuccessfully)
+            {
+                DisableReceive();
+                ReceiveComplete(0);
+                _ = AwaitFlush(flushTask);
+                return;
+            }
+
+            async Task AwaitFlush(ValueTask<FlushResult> ft)
+            {
+                await ft;
+                EnableReceive();
+                ReceiveComplete(length);
+            }
+        }
+
+        private uint HandleEventPeerRecvAbort()
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendAbort()
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleStartComplete()
+        {
+            _started = StartState.Finished;
+            _sendResettableCompletionSource.Complete(MsQuicConstants.Success);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventSendShutdownComplete(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownComplete()
+        {
+            // TODO use another cts here.
+            _sendResettableCompletionSource.Complete(MsQuicConstants.Success);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventPeerSendClose()
+        {
+            _requestPipe.Writer.Complete();
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventSendComplete(ref MsQuicNativeMethods.StreamEvent evt)
+        {
+            _sendHandle.Free();
+            foreach (MemoryHandle gchBufferArray in _bufferArrays)
+            {
+                gchBufferArray.Dispose();
+            }
+            // TODO throw if a write failed?
+            uint errorCode = evt.Data.SendComplete.Canceled;
+            _sendResettableCompletionSource.Complete(MsQuicConstants.Success);
+
+            return MsQuicConstants.Success;
+        }
+
+        private void SetCallbackHandler()
+        {
+            _handle = GCHandle.Alloc(this);
+
+            _callback = new StreamCallbackDelegate(NativeCallbackHandler);
+            _api.SetCallbackHandlerDelegate(
+                _ptr,
+                _callback,
+                GCHandle.ToIntPtr(_handle));
+        }
+
+        // TODO this probably can be ReadOnlyMemory
+        public unsafe ValueTask<uint> SendAsync(
+           ReadOnlySequence<byte> buffers,
+           QUIC_SEND_FLAG flags)
+        {
+            int bufferCount = 0;
+            foreach (ReadOnlyMemory<byte> memory in buffers)
+            {
+                bufferCount++;
+            }
+
+            var quicBufferArray = new QuicBuffer[bufferCount];
+            _bufferArrays = new MemoryHandle[bufferCount];
+
```

Comment:
Why are you making new allocations every single send? I don't quite understand what's going on here.

Reply:
Can this buffer array (not buffers themselves) be allocated on the stack or does it need to last the lifetime of the async operation?

Reply:
Anything passed to MsQuic needs to survive until the send has completed.